lib/config/presets/index.ts

```diff
@@ -384,11 +407,48 @@ export async function resolveConfigPresets(
         val as RenovateConfig,
         baseConfig,
         ignorePresets,
-        existingPresets
+        existingPresets,
+        shallowResolve
       );
     }
   }
   logger.trace({ config: inputConfig }, 'Input config');
   logger.trace({ config }, 'Resolved config');
+  if (shallowResolve) {
+    handleExtendsArray(config, unresolvedPreserts);
+  }
   return config;
 }
+
+// handleExtendsArray:
+// This function edits extends array presets when shallowResolve is enabled
+// 1. remove resolved external presets from the extends array.
+// 2. clean duplicate presets keys (from different resolution levels).
+function handleExtendsArray(
+  config: AllConfig,
+  unresolvedPreserts: string[]
+): void {
+  if (config.extends === undefined || config.extends.length === 0) {
+    return;
+  }
+  config.extends = config.extends?.filter((e) => !shouldShallowResolve(e));
+  config.extends?.push(...unresolvedPreserts);
+  const uniqueExtends = new Set(config.extends?.values());
+  config.extends = Array.from(uniqueExtends);
+  if (config?.extends?.length === 0) {
+    // clean empty extends array
+    delete config.extends;
+  }
```

Comment:
```suggestion
  const presets = new Set([ ...config.extends.filter((e) => !shouldShallowResolve(e)), ...unresolvedPresets]);
  config.extends = Array.from(uniqueExtends);
  if (config.extends.length === 0) {
    // clean empty extends array
    delete config.extends;
  }
```

TypeScript should be smart enough to validate that `config.extends` isn't `undefined` after check and assignment